File: src/main/java/org/flighthub/Repository/TicketRepository.java
package org.flighthub.Repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.flighthub.Domain.Client;
import org.flighthub.Domain.Flight;
import org.flighthub.Domain.Ticket;
import org.flighthub.Domain.Tourist;
import org.flighthub.Utils.JdbcUtils;
import org.flighthub.Repository.*;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.*;
import org.flighthub.Domain.Tourist;
public class TicketRepository implements ITicketRepository {
    private final JdbcUtils JDBCconnection;

    private static final Logger logger = LogManager.getLogger();

    public TicketRepository(JdbcUtils connection) {
        this.JDBCconnection = connection;
    }

    @Override
    public Optional<Ticket> findOne(UUID id) {
        try {
            logger.traceEntry();
            logger.info("trying to find one Ticket");
            Connection connection = JDBCconnection.getConnection();
            String query = "SELECT * FROM Ticket WHERE id = ?";
            PreparedStatement statement = connection.prepareStatement(query);
            statement.setString(1, id.toString());
            ResultSet resultSet = statement.executeQuery();

            if (resultSet.next()) {
                ClientRepository clientRepository = new ClientRepository(JDBCconnection);
                Optional<Client> client = clientRepository.findOne(UUID.fromString(resultSet.getString("clientId")));
                Client client1 = client.orElse(null);
                FlightRepository flightRepository = new FlightRepository(JDBCconnection);
                Optional<Flight> flight = flightRepository.findOne(UUID.fromString(resultSet.getString("flightId")));
                Flight flight1 = flight.orElse(null);

                // Extrageți lista de id-uri ale turiștilor din tabela de asociere Ticket_Tourist
                String queryTourists = "SELECT tourist_id FROM Ticket_Tourist WHERE ticket_id = ?";
                PreparedStatement statementTourists = connection.prepareStatement(queryTourists);
                statementTourists.setString(1, id.toString());
                ResultSet resultSetTourists = statementTourists.executeQuery();

                List<Tourist> touristList=new ArrayList<>();
                while (resultSetTourists.next()) {
                    // Pentru fiecare id de turist, găsiți numele corespunzător în tabela Tourist
                    String queryTouristName = "SELECT name FROM Tourist WHERE id = ?";
                    PreparedStatement statementTouristName = connection.prepareStatement(queryTouristName);
                    statementTouristName.setString(1, resultSetTourists.getString("tourist_id"));
                    ResultSet resultSetTouristName = statementTouristName.executeQuery();

                    if (resultSetTouristName.next()) {
                        Tourist tourist=new Tourist(resultSetTouristName.getString("name"));
                        tourist.setId(UUID.fromString(queryTourists));
                        touristList.add(tourist);
                    }
                }


                int seats = resultSet.getInt("seats");
                Ticket ticket = new Ticket(client1, touristList, flight1, seats);
                ticket.setId(UUID.fromString(resultSet.getString("id")));
                return Optional.of(ticket);
            }
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }

        return Optional.empty();
    }


    @Override
    public Iterable<Ticket> findAll() {
        List<Ticket> tickets = new ArrayList<>();

        try {
            logger.traceEntry();
            logger.info("trying to find multiple Tickets");
            Connection connection = JDBCconnection.getConnection();
            String query = "SELECT * FROM Ticket";
            PreparedStatement statement = connection.prepareStatement(query);
            ResultSet resultSet = statement.executeQuery();

            while (resultSet.next()) {
                ClientRepository clientRepository = new ClientRepository(JDBCconnection);
                Optional<Client> client = clientRepository.findOne(UUID.fromString(resultSet.getString("clientId")));
                Client client1 = client.orElse(null);
                FlightRepository flightRepository = new FlightRepository(JDBCconnection);
                Optional<Flight> flight = flightRepository.findOne(UUID.fromString(resultSet.getString("flightId")));
                Flight flight1 = flight.orElse(null);
                /// Extrageți lista de id-uri ale turiștilor din tabela de asociere Ticket_Tourist
                String queryTourists = "SELECT tourist_id FROM Ticket_Tourist WHERE ticket_id = ?";
                PreparedStatement statementTourists = connection.prepareStatement(queryTourists);
                statementTourists.setString(1, resultSet.getString("id"));
                ResultSet resultSetTourists = statementTourists.executeQuery();

                List<Tourist> touristList=new ArrayList<>();
                while (resultSetTourists.next()) {
                    // Pentru fiecare id de turist, găsiți numele corespunzător în tabela Tourist
                    String queryTouristName = "SELECT name FROM Tourist WHERE id = ?";
                    PreparedStatement statementTouristName = connection.prepareStatement(queryTouristName);
                    statementTouristName.setString(1, resultSetTourists.getString("tourist_id"));
                    ResultSet resultSetTouristName = statementTouristName.executeQuery();

                    if (resultSetTouristName.next()) {
                        Tourist tourist=new Tourist(resultSetTouristName.getString("name"));
                        //tourist.setId(UUID.fromString(queryTourists));
                        tourist.setId(UUID.fromString(resultSetTourists.getString("tourist_id")));
                        touristList.add(tourist);
                    }
                }

                int seats = resultSet.getInt("seats");
                Ticket ticket = new Ticket(client1, touristList, flight1, seats);
                ticket.setId(UUID.fromString(resultSet.getString("id")));
                tickets.add(ticket);
            }
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
            ;
        }

        return tickets;
    }

    @Override
    public Ticket save(Ticket entity) {
        try {
            logger.traceEntry();
            logger.info("trying to save one Ticket");
            Connection connection = JDBCconnection.getConnection();
            String query = "INSERT INTO Ticket (id, clientId, flightId, seats) VALUES (?, ?, ?, ?)";
            PreparedStatement statement = connection.prepareStatement(query);
            UUID ticketId = UUID.randomUUID(); // Generați un id unic pentru bilet
            statement.setString(1, ticketId.toString());
            statement.setString(2, entity.getClient().getId().toString());
            statement.setString(3, entity.getFlight().getId().toString());
            statement.setInt(4, entity.getSeats());
            entity.setId(ticketId); // Setează id-ul biletului cu id-ul generat
            statement.executeUpdate();

            // Salvare asociere turisti-bilet
            for (Tourist tourist : entity.getTouristList()) {
                String insertQuery = "INSERT INTO Ticket_Tourist (ticket_id, tourist_id) VALUES (?, ?)";
                PreparedStatement insertStatement = connection.prepareStatement(insertQuery);
                insertStatement.setString(1, ticketId.toString());
                insertStatement.setString(2, tourist.getId().toString());
                insertStatement.executeUpdate();
            }
            FlightRepository flightRepository = new FlightRepository(JDBCconnection);
            Optional<Flight> optionalFlight = flightRepository.findOne(entity.getFlight().getId());
            if (optionalFlight.isPresent()) {
                Flight flight = optionalFlight.get();
                int newAvailableSeats = flight.getAvailableSeats() - entity.getSeats();
                flightRepository.updateAvailableSeats(flight.getId(), newAvailableSeats);
            }
        } catch (SQLException e) {
            System.out.println("Error Repo " + e);
            logger.error(e);
        }

        return entity;
    }


    @Override
    public Optional<Ticket> delete(UUID id) {
        Optional<Ticket> ticketOptional = findOne(id);

        if (ticketOptional.isPresent()) {
            try {
                logger.traceEntry();
                logger.info("trying to delete one Ticket");
                Connection connection = JDBCconnection.getConnection();
                String query = "DELETE FROM Ticket WHERE id = ?";
                PreparedStatement statement = connection.prepareStatement(query);
                statement.setString(1, id.toString());
                statement.executeUpdate();
            } catch (SQLException e) {
                System.out.println("Error Repo " + e);
                logger.error(e);
            }
        }

        return ticketOptional;
    }

    @Override
    public Optional<Ticket> update(Ticket entity) {
        Optional<Ticket> ticketOptional = findOne(entity.getId());

        if (ticketOptional.isPresent()) {
            try {
                logger.traceEntry();
                logger.info("trying to update one Ticket");
                Connection connection = JDBCconnection.getConnection();
                String query = "UPDATE Ticket SET clientId = ?,  flightId = ?, seats = ? WHERE id = ?";
                PreparedStatement statement = connection.prepareStatement(query);
                statement.setString(1, entity.getClient().getId().toString());
                statement.setString(2, entity.getFlight().getId().toString());
                statement.setObject(3, entity.getSeats());
                statement.setString(4, entity.getId().toString());
                statement.executeUpdate();
            } catch (SQLException e) {
                System.out.println("Error Repo " + e);
                logger.error(e);
            }
        }

        return ticketOptional;
    }
}

File: Server - C#/persistence/TicketRepository.cs
using FlightHubC_.Domain;
using FlightHubC_.Utils;
//using log4net;
using System;
using System.Collections.Generic;
using System.Data.SqlClient;

namespace FlightHubC_.Repository
{
    public class TicketRepository : ITicketRepository
    {
        private readonly SqlUtils sqlUtils;
        //private static readonly log4net.ILog log = LogManager.GetLogger(typeof(Program));

        public TicketRepository(SqlUtils sqlUtils)
        {
            this.sqlUtils = sqlUtils;
        }

        public Ticket FindOne(Guid id)
        {
            string query = "SELECT * FROM Ticket WHERE Id = @Id";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                //log.Info("finding one Ticket");
                connection.Open();
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    command.Parameters.AddWithValue("@Id", id);
                    using (SqlDataReader reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            ClientRepository clientRepository = new ClientRepository(sqlUtils);
                            Client client = clientRepository.FindOne((Guid)reader["clientId"]);
                            FlightRepository flightRepository = new FlightRepository(sqlUtils);
                            Flight flight = flightRepository.FindOne((Guid)reader["flightId"]);
                            int seats = (int)reader["seats"];

                            // Extrageți lista de id-uri ale turiștilor din tabela de asociere Ticket_Tourist
                            List<Guid> touristIds = new List<Guid>();
                            string queryTourists = "SELECT tourist_id FROM Ticket_Tourist WHERE ticket_id = @Id";
                            using (SqlCommand commandTourists = new SqlCommand(queryTourists, connection))
                            {
                                commandTourists.Parameters.AddWithValue("@Id", id);
                                using (SqlDataReader readerTourists = commandTourists.ExecuteReader())
                                {
                                    while (readerTourists.Read())
                                    {
                                        touristIds.Add((Guid)readerTourists["tourist_id"]);
                                    }
                                }
                            }

                            List<Tourist> tourists = new List<Tourist>();
                            foreach (var touristId in touristIds)
                            {
                                string queryTouristName = "SELECT name FROM Tourist WHERE id = @TouristId";
                                using (SqlCommand commandTouristName = new SqlCommand(queryTouristName, connection))
                                {
                                    commandTouristName.Parameters.AddWithValue("@TouristId", touristId);
                                    using (SqlDataReader readerTouristName = commandTouristName.ExecuteReader())
                                    {
                                        if (readerTouristName.Read())
                                        {
                                            Tourist tourist = new Tourist((string)readerTouristName["name"]);
                                            tourist.Id = touristId;
                                            tourists.Add(tourist);
                                        }
                                    }
                                }
                            }

                            Ticket ticket = new Ticket(client, tourists, flight, seats);
                            ticket.Id = id;
                            return ticket;
                        }
                    }
                }
            }
            return null;
        }

        public IEnumerable<Ticket> FindAll()
        {
            List<Ticket> tickets = new List<Ticket>();
            string query = "SELECT * FROM Ticket ";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                //log.Info("finding all Tickets");
                connection.Open();
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    using (SqlDataReader reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            ClientRepository clientRepository = new ClientRepository(sqlUtils);
                            Client client = clientRepository.FindOne((Guid)reader["clientId"]);
                            Guid Id = reader.GetGuid(reader.GetOrdinal("id"));
                            String[] TouristsName = reader["touristsName"].ToString().Split(',');
                            FlightRepository flightRepository = new FlightRepository(sqlUtils);
                            Flight flight = flightRepository.FindOne(Guid.Parse(reader["flightId"].ToString()));
                            int seats = (int)reader["seats"];
                            // Extrageți lista de id-uri ale turiștilor din tabela de asociere Ticket_Tourist
                            List<Guid> touristIds = new List<Guid>();
                            string queryTourists = "SELECT tourist_id FROM Ticket_Tourist WHERE ticket_id = @Id";
                            using (SqlCommand commandTourists = new SqlCommand(queryTourists, connection))
                            {
                                commandTourists.Parameters.AddWithValue("@Id", Id);
                                using (SqlDataReader readerTourists = commandTourists.ExecuteReader())
                                {
                                    while (readerTourists.Read())
                                    {
                                        touristIds.Add((Guid)readerTourists["tourist_id"]);
                                    }
                                }
                            }

                            List<Tourist> tourists = new List<Tourist>();
                            foreach (var touristId in touristIds)
                            {
                                string queryTouristName = "SELECT name FROM Tourist WHERE id = @TouristId";
                                using (SqlCommand commandTouristName = new SqlCommand(queryTouristName, connection))
                                {
                                    commandTouristName.Parameters.AddWithValue("@TouristId", touristId);
                                    using (SqlDataReader readerTouristName = commandTouristName.ExecuteReader())
                                    {
                                        if (readerTouristName.Read())
                                        {
                                            Tourist tourist = new Tourist((string)readerTouristName["name"]);
                                            tourist.Id = touristId;
                                            tourists.Add(tourist);
                                        }
                                    }
                                }
                            }

                            Ticket ticket = new Ticket(client, tourists, flight, seats);
                            ticket.Id = Id;
                            tickets.Add(ticket);
                        }
                    }
                }
            }
            return tickets;
        }

        public Ticket Save(Ticket entity)
        {
            string query = "INSERT INTO Ticket (id, clientId, flightId, seats) VALUES (@Id, @ClientId, @FlightId, @Seats)";

            using (SqlConnection connection = sqlUtils.GetConnection())
            {
                //log.Info("saving one Ticket");

                connection.Open();
                using (SqlCommand command = new SqlCommand(query, connection))
                {
                    Guid ticketId = Guid.NewGuid(); // Generați un id unic pentru bilet
                    entity.Id = ticketId;
                    command.Parameters.AddWithValue("@Id", ticketId);
                    command.Parameters.AddWithValue("@ClientId", entity.client.Id);
                    command.Parameters.AddWithValue("@FlightId", entity.flight.Id);
                    command.Parameters.AddWithValue("@Seats", entity.seats);
                    command.ExecuteNonQuery();
                }

                // Salvare asociere turisti-bilet
                foreach (Tourist tourist in entity.tourists)
                {
                    string insertQuery = "INSERT INTO Ticket_Tourist (ticket_id, tourist_id) VALUES (@TicketId, @TouristId)";
                    using (SqlCommand insertCommand = new SqlCommand(insertQuery, connection))
                    {
                        insertCommand.Parameters.AddWithValue("@TicketId", entity.Id);
                        insertCommand.Parameters.AddWithValue("@TouristId", tourist.Id);
                        insertCommand.ExecuteNonQuery();
                    }
                }

                // Actualizare disponibilitate locuri în zbor
                FlightRepository flightRepository = new FlightRepository(sqlUtils);
                Flight flight = flightRepository.FindOne(entity.flight.Id);
                if (flight != null)
                {
                    int newAvailableSeats = flight.AvailableSeats - entity.seats;
                    flightRepository.updateAvailableSeats(flight.Id, newAvailableSeats);
                }
            }

            return entity;
        }


        public Ticket Delete(Guid id)
        {
            Ticket ticket = FindOne(id);
            if (ticket != null)
            {
                string query = "DELETE FROM Ticket WHERE id = @Id";
                using (SqlConnection connection = sqlUtils.GetConnection())
                {
                    connection.Open();
                    using (SqlCommand command = new SqlCommand(query, connection))
                    {
                        command.Parameters.AddWithValue("@Id", id);
                        command.ExecuteNonQuery();
                    }
                }
            }
            return ticket;
        }

        public Ticket Update(Ticket entity)
        {

            return entity;
        }
    }
}
